Interaction functions

joshkel opened the pull request:
For discussion / work in progress - cleaning up and expanding the interaction modes API, as discussed in #9961.

Specific changes:

* Consolidate `getRelativePosition`. See c7659383 for details. I added a test to verify this; it depends on a local, not-yet-submitted change in chartjs-test-utils' `triggerMouseEvent` to return the mouse event that it triggers. I'll open a separate PR for that if this change seems reasonable.
* Add a new public `Chart` method, `isPointInArea`, to refactor and publicize logic that was duplicated in core.controller.js and core.interactions.js.
* Change `getAxisItems` to have an interface consistent with other `get...Items` functions.
* Consolidate `evaluateAllVisibleItems` and `optimizedEvaluateItems` - `optimizeEvaluateItems` falls back to evaluating everything if the optimization fails, and I couldn't think of a reason why `getAxisItems` wouldn't want to benefit from the optimized binary search if it could.

Proposed API:

1. `getRelativePosition` is already exported by `'chart.js/helpers'` and, after these changes, can efficiently be used as is.
2. Add a new public `Chart` method, `isPointInArea`.
3. Rename `optimizedEvaluateItems` to `evaluateInteractionItems` and export it.
4. _Optional_: Export the individual `get...Items` functions: `getIntersectItems`, `getNearestItems`, `getAxisItems`. (I went ahead and marked these and `evaluateInteractionItems` as `export` in the current PR.)
5. _Optional_: Export `getDistanceMetricForAxis`.

I'm currently thinking that the `get...Items` and `getDistanceMetricForAxis` functions would _not_ be exported: They're straightforward to implement using existing public APIs, and they may need to vary due to specific interaction needs, so the costs of freezing them as part of the public API may not outweigh the benefits.

For reference, if I were to export some sort of `get...Items` functionality, here's what I'm using for my own code. It's based off of `getNearestItems`, modified to take a callback: (This function could be exported, and `getNearestCartesianItems` could be changed to use this.)

```
function evaluateNearestItems(
  callback: (
    element: Element & VisualElement,
    datasetIndex: number,
    index: number,
    distance: number
  ) => void,
  chart: Chart,
  position: Point,
  axis: Axis,
  intersect?: boolean,
  useFinalPosition?: boolean
) {
  const distanceMetric = getDistanceMetricForAxis(axis);

  const minPadding = (chart as Chart & { _minPadding: number })._minPadding;
  if (!_isPointInArea(position, chart.chartArea, minPadding)) {
    return false;
  }

  const evaluationFunc: EvaluateHandler = (element, datasetIndex, index) => {
    if (intersect && !element.inRange(position.x, position.y, useFinalPosition)) {
      return;
    }

    const center = element.getCenterPoint(useFinalPosition);
    if (
      !_isPointInArea(center, chart.chartArea, minPadding) &&
      !element.inRange(position.x, position.y, useFinalPosition)
    ) {
      return;
    }
    const distance = distanceMetric(position, center);
    callback(element, datasetIndex, index, distance);
  };

  optimizedEvaluateItems(chart, axis, position, evaluationFunc);
  return true;
}
```

kurkle commented:
> Optional: Export the individual get...Items functions: getIntersectItems, getNearestItems, getAxisItems. (I went ahead and marked these and evaluateInteractionItems as export in the current PR.)

Would it actually be enough to only export these "higher level" functions? Would probably leave more room for internal changes.

joshkel commented:
@kurkle For my purposes, I don't think that would help. Going off of the two motivating examples from #9961 (`nearestPerDataset`, and nearest but with some sort of `intersectRadius` or `maxDistance` limit):

* I don't see a good way to implement `nearestPerDataset` using the higher-level `get...Items` functions; it would need `optimizedEvaluateItems` / `evaluateInteractionItems` in order to see all of the candidates, so that it can evaluate per dataset.
* nearest-with-limit could be implemented using `getNearestItems`, although it would then have to duplicate `getNearestItems`' distance calculation to decide whether items are too close.

(I realize that there's [discussion of other changes](https://github.com/chartjs/Chart.js/issues/9961#issuecomment-986946446) that might remove the need for either or both of these specific examples, but hopefully the examples can still demonstrate what might make for a more broadly useful API.)

kurkle commented:
I don't have any objections to this.

kurkle commented:
> ... chartjs-test-utils' triggerMouseEvent to return the mouse event that it triggers. I'll open a separate PR for that if this change seems reasonable.

There is a change pending release on the test-utils, could also include your changes if you open the PR.

joshkel commented:
@kurkle Submitted as https://github.com/chartjs/chartjs-test-utils/pull/19. Thanks for the reminder.

For this PR: I listed some optional functions for export as items 4 and 5 (the individual `get...Items` functions: `getIntersectItems`, `getNearestItems`, `getAxisItems` - and `getDistanceMetricForAxis `) Do you have an opinion / preference on whether those should be included? I can go ahead and finalize the PR based on that decision.

Thank you.

etimberg commented on src/helpers/helpers.dom.js:
```diff
@@ -59,8 +59,7 @@ function getPositionedStyle(styles, style, suffix) {
 const useOffsetPos = (x, y, target) => (x > 0 || y > 0) && (!target || !target.shadowRoot);
 
-function getCanvasPosition(evt, canvas) {
-  const e = evt.native || evt;
```
What's the reason for no longer preferring the native event?

joshkel replied:
My understanding the code that's currently in master:

* `getCanvasPosition` prefers the native event, like you said - in other words, it ensures that, regardless of whether it's given a `ChartEvent` or a native `Event`, it starts its operations from the native x, y coordinates.
* `getCanvasPosition` is only used by helpers.dom.js's version of `getRelativePosition`. To summarize  and maybe oversimplify, it starts from the native x, y coordinates and calculates the x, y coordinates relative to the chart.
* core.interaction.js's version of `getRelativePosition` to check if it's given a `ChartEvent` and, if so, to return the x, y coordinates from the `ChartEvent` - in other words, if it's given a `ChartEvent` instead of a native `Event`, it assumes / treats the `ChartEvent`'s x, y as already giving the x, y coordinates relative to the chart.

So helpers.dom.js's `getRelativePosition` supports being given both a `ChartEvent` and a native `Event`, and if it's given a `ChartEvent`, it ignores the `ChartEvent`'s relative x, y coordinates and (re)calculates the relative coordinates starting from the native `Event`. To consolidate code and make it easier to export, I'm changing helpers.dom.js to have the same "use `ChartEvent`'s x, y" logic as core.interaction.js's version; after that change, `getCanvasPosition` is only ever given a native `Event`, so it no longer needs to check if it's given a `ChartEvent`.

I added some JSDoc types to help make this clearer. To keep the lint-tsc check happy, I had to add a `@ts-ignore`, because the current `getCanvasPosition` code relies on some loose conversions between `Event`, `TouchEvent`, and `Touch` that would require several changes to make TypeScript happy. I'm not sure if that fits with Chart.js's coding standards; if you want me to back that out or clean up the type handling to remove `@ts-ignore`, please let me know.

If I've misunderstood any of the above - i.e., if `ChartEvent'`s x, y coordinates are ever not relative to the chart - then this won't work. The helpers.dom.tests.js test is to help ensure that this is valid.

etimberg replied:
You're understanding that the `ChartEvent` has the relative position is correct and re-using that data makes sense.

I think the `ts-ignore` is fine for now. Some of these helper functions could be converted directly to Typescript in the future and we could remove the ignores then

etimberg commented on test/specs/helpers.dom.tests.js:
```diff
@@ -427,5 +427,29 @@ describe('DOM helpers tests', function() {
+      const nativeEvent = await jasmine.triggerMouseEvent(chart, 'mousemove', point);
+      const nativePosition = Chart.helpers.getRelativePosition(nativeEvent, chart);
+
+      expect(nativePosition).toEqual(chartPosition);
```
can you add one more expect here so that we know `chartPosition !== null`. Just want to confirm the `onHover` callback actually ran in case the `nativeEvent` is also somehow null

joshkel replied:
Done.

etimberg reviewed (commented):
Just a few questions / comments. Overall I like the direction this is going

kurkle commented on src/core/core.interaction.js:
```diff
@@ -121,15 +84,15 @@ function getDistanceMetricForAxis(axis) {
  * Helper function to get the items that intersect the event position
  * @param {Chart} chart - the chart
- * @param {object} position - the point to be nearest to
+ * @param {Point} position - the point to be nearest to
```
Could add the ", in relative coordinates" to all relevant functions

kurkle commented on src/helpers/helpers.dom.js:
```diff
@@ -79,6 +84,13 @@ function getCanvasPosition(evt, canvas) {
 
 export function getRelativePosition(evt, chart) {
+  if ('native' in evt) {
+    return {
```
Would maybe be a little faster to just return the `evt` here. @etimberg any objections to that?

etimberg replied:
no objections

kurkle reviewed (dismissed):
Couple of nits

kurkle commented:
> TypeError: Cannot use 'in' operator to search for 'native' in undefined

joshkel commented:
> TypeError: Cannot use 'in' operator to search for 'native' in undefined

@kurkle Are you seeing this error? If so, where?

joshkel commented:
@kurkle and @etimberg , please see my latest commits:

* I responded to the code review feedback. Thanks!
* I updated how the exports are done - including within the default export seems to be Chart.js's convention, instead of exporting directly, and exporting directly wouldn't make the files accessible via `import { something } from 'chart.js'`.
* I updated the docs and the TypeScript type definitions.
* For now, I'm *not* exporting the individual `get...Items` functions. I'm not sure that they're flexible enough for custom interaction modes (e.g., I haven't used them myself); based on discussions about not wanting to commit to a broader API than necessary, I think it's more useful to start with justing export the building blocks (`evaluateInteractionItems`, `Chart.isPointInArea`, and the updated `getRelativePosition`). However, I'm happy to add them back if you like.

kurkle commented:
> > TypeError: Cannot use 'in' operator to search for 'native' in undefined
> 
> @kurkle Are you seeing this error? If so, where?

CI fails with that. I guess the event can be undefined in some case.

tomaszkrym commented:
@kurkle looking at CI history ( https://github.com/chartjs/Chart.js/actions?query=workflow%3ACI+branch%3Amaster ) the tests seem to not be 100% deterministic, they sometimes fail randomly on master. The changes submitted by @joshkel (thank you for PR!) don't seem to touch code that failed tests in CI. Is it possible to rerun tests?

I would love to see this PR merged and released as we also need the "nearestPerDataset" mode in our project.

kurkle commented on src/helpers/helpers.dom.js:
```diff
@@ -78,7 +91,17 @@ function getCanvasPosition(evt, canvas) {
+ * @returns {{x: number, y: number}}
+ */
 export function getRelativePosition(evt, chart) {
+  if ('native' in evt) {
```
```suggestion
  if (evt && ('native' in evt)) {
```

kurkle commented on src/helpers/helpers.dom.js:
```diff
@@ -78,7 +91,17 @@ function getCanvasPosition(evt, canvas) {
+ * @returns {{x: number, y: number}}
+ */
 export function getRelativePosition(evt, chart) {
+  if (evt && ('native' in evt)) {
```
```suggestion
  if ('native' in evt) {
```

joshkel commented:
@kurkle Done. Thanks for following up on this, and sorry for the delay.

etimberg commented:
@kurkle am I OK to merge this?
